libraries/lib-snapping/SnapUtils.cpp

```diff
@@ -0,0 +1,315 @@
+   const Identifier& id, SnapConfig config, double time, bool nearest)
+{
+   auto item = Find(id);
+
+   if (item == nullptr)
+      return SnapResult { time, false };
+
+   return item->snapFunction(config, time, nearest);
+}
+
+SnapRegistryGroup::~SnapRegistryGroup()
+{
+}
+
+bool SnapRegistryGroup::Transparent() const
+{
+   return transparent;
+}
+
+SnapRegistryItem::SnapRegistryItem(
+   const Identifier& internalName, const TranslatableString& _label,
+   SnapFunctor _snapFunction)
+    : SingleItem { internalName }
+    , label { _label }
+    , snapFunction { _snapFunction }
+{
+}
+
+SnapRegistryItem::~SnapRegistryItem()
+{
+}
+
+Registry::BaseItemPtr SnapFunction(
+   const Identifier& functionId, const TranslatableString& label,
+   SnapFunctor functor)
+{
+   return std::make_unique<SnapRegistryItem>(
+      functionId, label, std::move(functor));
+}
+
+SnapRegistryItemRegistrator::SnapRegistryItemRegistrator(
+   const Registry::Placement& placement, Registry::BaseItemPtr pItem)
+    : RegisteredItem { std::move(pItem), placement }
+{
+}
+
+namespace
+{
+SnapResult SnapWithMultiplier (double value, double multiplier, bool nearest)
+{
+   auto result = nearest ? std::round(value * multiplier) / multiplier :
+                           std::floor(value * multiplier) / multiplier;
+
+   return SnapResult { result, true };
+}
+
+
+SnapFunctor SnapToTime(double multiplier) {
+   return [multiplier](SnapConfig, double value, bool nearest)
+   { return SnapWithMultiplier(value, multiplier, nearest); };
+}
+
+SnapResult SnapToSamples (SnapConfig cfg, double value, bool nearest)
+{
+   return SnapWithMultiplier(value, cfg.rate, nearest);
+}
+
+SnapFunctor SnapToFrames(double fps)
+{
+   return [fps](SnapConfig, double value, bool nearest)
+   { return SnapWithMultiplier(value, fps, nearest); };
+}
+
+/*
+   bps = tempo / 60
+   1/lower takes 1/bps
+*/
+
+SnapFunctor SnapToBar ()
+{
+   return [](SnapConfig cfg, double value, bool nearest)
+   {
+      const auto multiplier = (cfg.tempo / 60.0) / cfg.timeSignature.first;
+      return SnapWithMultiplier(value, multiplier, nearest);
+   };
+}
+
+SnapFunctor SnapToBeat(int divisor)
+{
+   return [divisor](SnapConfig cfg, double value, bool nearest)
+   {
+      const auto multiplier =
+         (cfg.tempo / 60.0) * divisor / cfg.timeSignature.second;
+      return SnapWithMultiplier(value, multiplier, nearest);
+   };
+}
+
+SnapFunctor SnapToTriplets(int divisor)
+{
+   return [divisor](SnapConfig cfg, double value, bool nearest)
+   {
+      const auto tripletDivisor = 3 * (divisor / 2);
+      const auto multiplier =
+         (cfg.tempo / 60.0) * tripletDivisor / cfg.timeSignature.second;
+      return SnapWithMultiplier(value, multiplier, nearest);
+   };
+}
+
+SnapRegistryItemRegistrator beats {
+   Registry::Placement { {}, { Registry::OrderingHint::Begin } },
+   SnapFunctionGroup(
+      "beats", XO("Beats"), true, SnapFunction("bar", XO("Bar"), SnapToBar()),
```

Comment:
The transparency of the groups doesn't matter unless they are nameless

Reply:
Potentially we will change the transparency or even make it dependent on user preferences.

Reply:
On the other hand I see you reuse transparency explicitly for another purpose.  I was only thinking of how Registry.cpp assigns meaning to transparency in the merging procedure.